### ISSUE TITLE
Output to spreadsheet without filling in row headers until header name changes?

### ISSUE THREAD

**Issue by jb261** [aiguofer/gspread-pandas#91]
Starting with the following dictionary:

```
    test_dict = {'header1_1': {'header2_1': {'header3_1': {'header4_1': ['322.5', 330.0, -0.28],
        'header4_2': ['322.5', 332.5, -0.26]},
       'header3_2': {'header4_1': ['285.0', 277.5, -0.09],
        'header4_2': ['287.5', 277.5, -0.12]}},
      'header2_2': {'header3_1': {'header4_1': ['345.0', 357.5, -0.14],
        'header4_2': ['345.0', 362.5, -0.14]},
       'header3_2': {'header4_1': ['257.5', 245.0, -0.1],
        'header4_2': ['257.5', 240.0, -0.08]}}}}
```
I want the headers in the index, so I reform the dictionary:

```
    reformed_dict = {}
    for outerKey, innerDict in test_dict.items():
        for innerKey, innerDict2 in innerDict.items():
            for innerKey2, innerDict3 in innerDict2.items():
                for innerKey3, values in innerDict3.items():
                    reformed_dict[(outerKey,
                            innerKey, innerKey2, innerKey3)] = values
```
And assign column names to the headers:

```
    keys = reformed_dict.keys()
    values = reformed_dict.values()
    index = pd.MultiIndex.from_tuples(keys, names=["H1", "H2", "H3", "H4"])
    df = pd.DataFrame(data=values, index=index)

```
That gets to a dataframe that looks like this:
![vDQ5R](https://user-images.githubusercontent.com/52250987/204066629-49523ae9-8b6a-4cb1-a744-1241bcc85af5.png)

Issue:  When I write this dataframe to Google Sheets using [gspread-pandas][2]:

```
    s.open_sheet('test')
    Spread.df_to_sheet(s, df, index=True, headers=True, start='A8', replace=False) 

```
The table looks like this: 
![midf_1](https://user-images.githubusercontent.com/52250987/204067009-584ea690-1814-4e8c-97fa-73574507ebb7.png)

Would it be possible to manipulate the dataframe to get written to spreadsheet like this:
![midf_2](https://user-images.githubusercontent.com/52250987/204067061-55f88ded-06ff-488c-a770-4c43c659cfcf.png)

  [1]: https://i.stack.imgur.com/vDQ5R.png
  [2]: https://gspread-pandas.readthedocs.io/en/latest/index.html

**Comment by aiguofer** [comment]
Unfortunately this isn't possible right now. There is an option to do this by merging cells in the headers by passing `merge_headers=True`. I feel like we could try a similar thing to do `merge_indexes=True`, although this would merge the cells instead of leaving them blank. Thoughts?

**Comment by jb261** [comment]
It's a bit over my head, but maybe the kind of option I'm looking for is best implemented in pandas.

As for the idea of merging cells, seems that could work in terms of desired output, at least visually. But depending on how users want to use the data once it's written to spreadsheet, that might not be helpful. If users want to manipulate the data or use it as the basis of queries, filters, etc., the merged cells could be a problem.

**Comment by aiguofer** [comment]
Well if you want to create filters, queries, etc merging is the right thing to do. For example, if you write some function that should only apply to rows with `header1_1` then the way you showed wouldn't do the right thing because the only row with that value is `4`. However, if `A4-A11` are merged then the function would do the right thing since rows `4-11` would all have that value.

**Comment by jb261** [comment]
If you want to try to implement a `merge_indexes`, I'm happy to test the beta.

**Comment by aiguofer** [comment]
@jb261 great, I just pushed #92 which adds that... I tested using the df you provided and it seemed to work, let me know if this lets you do what you're hoping for!

**Comment by aiguofer** [comment]
Hmm... based on [this SO post](https://webapps.stackexchange.com/questions/110277/how-do-i-reference-the-values-of-merged-cells-in-formulas) it seems like merged cells act like blank cells, so I was wrong regarding formulas and such.

The one benefit of the merged cells over blank cells is that if you read the sheet back into a DF it will look up the correct value for merged cells (I had to implement that).

**Comment by jb261** [comment]
> @jb261 great, I just pushed #92 which adds that... I tested using the df you provided and it seemed to work, let me know if this lets you do what you're hoping for!

Just had a chance to try this out - it works great!

Yeah, I can see the merged cells acting as empty cells, but I think this `merge_index=True` is useful when dealing with large hierarchical datasets, where there will typically lots of repetition in some of the indexes. Makes it easier to read/visualize. And your modification to be able to read this back into a dataframe and pick up the correct value is a good idea. Thanks!

**Comment by jb261** [comment]
Once cells are merged in Google Sheets, if you rerun Spread.df_to_sheet with merge_index=False, the cells stay merged, so as of right now, this is a one way street. Not an issue in my particular use case, but probably not the result typical user would expect.

**Comment by aiguofer** [comment]
Thanks for pointing that out! I might do a little more testing to ensure at least the merged indexes update correctly on a second run with an updated DF

**Comment by jb261** [comment]
> Thanks for pointing that out! I might do a little more testing to ensure at least the merged indexes update correctly on a second run with an updated DF

Yeah, I ran into this issue today. The target spreadsheet had previously had a dataframe written to it, so a couple of columns were merged. Updating that had been going fine because each update had the same number of rows corresponding to headers in the merged column indexes. Today's update had fewer rows for each merged column index, so `Spread.df_to_sheet` generated an error, along the lines of it was unable to unmerge the cells. I manually unmerged the merged columns in the spreadsheet and was able to run `Spread.df_to_sheet` without issue. So as of right now, user has to manually unmerge if the dataframe rows don't match up, in terms of merged column indexe, with what has previously been written to spreadsheet.

**Comment by aiguofer** [comment]
I just pushed some changes to reset the merged cells before making the request to merge again... this should solve that.

One thing to note is that when you're updating the data with less rows, the rest of the rows that were previously there are not cleared out. I ended up adding a `replace` flag to `df_to_sheet` to ensure there were no issues with lingering data. However, that does resize the spreadsheet to only the cells needed to fit the data.

**Comment by jb261** [comment]
Just tried and yeah, the update to merge index functionality is working.

> One thing to note is that when you're updating the data with less rows, the rest of the rows that were previously there are not cleared out. I ended up adding a `replace` flag to `df_to_sheet` to ensure there were no issues with lingering data. However, that does resize the spreadsheet to only the cells needed to fit the data.

Wasn't the `replace` parameter already part of `df_to_sheet`? I generally set `replace=False` because I usually have other text or data in the worksheet, so I don't want the worksheet getting wiped.

**Comment by realPivo** [comment]
I encountered the same problem (wish) as the op. @aiguofer could you merge #92 to master?
I'm using it in Google Colab and it would be much easier to access through the pip package.

**Comment by aiguofer** [comment]
Hey sorry forgot to finish this earlier! merged and published in version 3.2.3
